libraries/AP_PacketDigitalCAN/AP_PacketDigitalCAN.cpp

```diff
@@ -0,0 +1,609 @@
+        break;
+
+    case Config_Mode::SET_VOUT:
+        if (_config.vout_is_set) {
+            break;
+        }
+        _config.vout_is_set = true;
+        for (uint8_t i=0; i<device_count; i++) {
+            send_command(PacketType::VOLTAGE_SET, MPPT_devices[i].serialnumber, _config.vout);
+        }
+        break;
+
+    case Config_Mode::SET_CVT:
+        if (_config.cvt_is_set) {
+            break;
+        }
+        _config.cvt_is_set = true;
+        for (uint8_t i=0; i<device_count; i++) {
+            send_command(PacketType::CVT_SET, MPPT_devices[i].serialnumber, _config.cvt);
+        }
+        break;
+
+    case Config_Mode::SET_ALGORITHM:
+        if (_config.algorithm_is_set) {
+            break;
+        }
+        _config.algorithm_is_set = true;
+        for (uint8_t i=0; i<device_count; i++) {
+            send_command(PacketType::ALGORITHM_SET, MPPT_devices[i].serialnumber, _config.algorithm);
+        }
+        break;
+    }
+}
+
+// parse inbound frames
+void AP_PacketDigitalCAN::handle_frame(const AP_HAL::CANFrame &frame)
+{
+    const uint16_t serialnumber = frame.id & 0x0000FFFF;
+    if (serialnumber == 0) {
+        // This is for broadcast and I don't think we should allow this inbound.
+        return;
+    }
+
+    WITH_SEMAPHORE(_sem);
+
+    uint8_t index = get_device_index(serialnumber);
+    if (index == UINT8_MAX) {
+        // we don't know this device
+        if (device_count >= device_count_max) {
+            // we don't have any room to remember it
+            return;
+        }
+        // add it
+        index = device_count;
+        device_count++;
+        MPPT_devices[index].serialnumber = serialnumber;
+        //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u New device", (unsigned)serialnumber);
+
+        send_command(PacketType::VOLTAGE_GET, serialnumber);
+        send_command(PacketType::ALGORITHM_GET, serialnumber);
+        send_command(PacketType::CVT_GET, serialnumber);
+    }
+
+    const uint8_t sequence = frame.data[0];
+    //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u Rx Seq %u -> %u", (unsigned)serialnumber, (unsigned)sequence_prev, (unsigned)sequence);
+
+    const bool bad_sequence = ((sequence_prev+1) != sequence);
+    if (bad_sequence) {
+        // we've dropped a packet somewhere
+        // TODO: handle lost packets
+        //gcs().send_text(MAV_SEVERITY_DEBUG, "PDCAN: %u Rx bad sequence ", (unsigned)serialnumber);
+    }
+
+    sequence_prev = sequence;
+
+    if (index >= device_count) {
+        // not sure how this can happen, but lets protect the array bounds just in case
+        return;
+    }
+
+    switch ((PacketType)frame.data[1]) {
+    case PacketType::STREAM_FAULT:
+        MPPT_devices[index].faults = (Fault_Flags)frame.data[2];
+        MPPT_devices[index].temperature = frame.data[3];
+        break;
+
+    case PacketType::STREAM_INPUT:
+        MPPT_devices[index].input.voltage = fixed2float(UINT16_VALUE(frame.data[2], frame.data[3]));
```

Comment:
It would be nice to use a structure to pull this packet apart rather than in-line like this.  But I don't know its really critical, so wait for someone else to comment.

Reply:
its only a couple, I think we can leave this as-is